---
title: "MDS analysis"
output: html_notebook
---

Goal: Do an MDS plot using Irene's SNPs to look for relatedness among plants


## Snp calling with freebayes


First, need to call SNPs using all BAM files at once:

working in directory `2019/IreneSnps` on whitney

first, sort bams

```{r, engine='bash', eval=FALSE}
for f in SNPanalysis/*LT*rmdup.bam
    do 
       newname=`basename $f .bam`_sort.bam
       samtools sort -o $newname --reference SNPanalysis/Pinsaporeference1 $f 
    done
```


next, assign read groups to bams

```{r, engine='bash', eval=FALSE}
input=""

for f in `ls *sort.bam`
   do
      rg=`basename $f _rmdup_sort.bam`
      input="$input -b $f -r $rg -s $rg"
  done

echo $input

bamaddrg $input > LT_rmdup_sort_combined.bam

samtools index LT_rmdup_sort_combined.bam
```


```{r, engine='bash', eval=FALSE}
freebayes -f SNPanalysis/Pinsaporeference1 --no-indels --no-mnps --no-complex LT_rmdup_sort_combined.bam > LT.vcf &
```

try parallel
```{r, engine='bash', eval=FALSE}
ulimit -n 4000
/usr/local/stow/freebayes/scripts/fasta_generate_regions.py Pinsaporeference1.fai 100000 > regions
./freebayes-parallel regions 8 -f Pinsaporeference1 --no-indels --no-mnps --no-complex LT_rmdup_sort_combined.bam > LT.vcf 
```

(note: I edited the freebayes-parallel script so that it would work...)

Takes about 12 hours

```{r, engine='bash', eval=FALSE}
scp whitney.plb.ucdavis.edu:2019/IreneSnps/LT.vcf.gz ../input/
```

## now analyze vcf in R

```{r}
library(tidyverse)
```

get the data
```{r}
snps <- read_tsv("../input/LT.vcf.gz", na = c("","NA","."),comment="#",col_names = FALSE)
head(snps)
```

get the header
```{r}
vcf.header <- system("zgrep '#C' ../input/LT.vcf.gz",intern = TRUE) 
vcf.header
vcf.header <- vcf.header %>% 
  str_replace("#","") %>% #get rid of the pound sign
  str_split(pattern = "\t") %>% #split on the tabs
  magrittr::extract2(1)
vcf.header
```

```{r}
colnames(snps) <- vcf.header
head(snps)
```



