---
title: "PCA analysis"
output: html_notebook
---

Goal: Do a PCA plot using Irene's SNPs to look for relatedness among plants.  Also figure out loading (which genes contribute most)


## Snp calling with freebayes


First, need to call SNPs using all BAM files at once:

working in directory `2019/IreneSnps` on whitney

first, sort bams

```{r, engine='bash', eval=FALSE}
for f in SNPanalysis/*LT*rmdup.bam
    do 
       newname=`basename $f .bam`_sort.bam
       samtools sort -o $newname --reference SNPanalysis/Pinsaporeference1 $f 
    done
```


next, assign read groups to bams

```{r, engine='bash', eval=FALSE}
input=""

for f in `ls *sort.bam`
   do
      rg=`basename $f _rmdup_sort.bam`
      input="$input -b $f -r $rg -s $rg"
  done

echo $input

bamaddrg $input > LT_rmdup_sort_combined.bam

samtools index LT_rmdup_sort_combined.bam
```


```{r, engine='bash', eval=FALSE}
freebayes -f SNPanalysis/Pinsaporeference1 --no-indels --no-mnps --no-complex LT_rmdup_sort_combined.bam > LT.vcf &
```

try parallel
```{r, engine='bash', eval=FALSE}
ulimit -n 4000
/usr/local/stow/freebayes/scripts/fasta_generate_regions.py Pinsaporeference1.fai 100000 > regions
./freebayes-parallel regions 8 -f Pinsaporeference1 --no-indels --no-mnps --no-complex LT_rmdup_sort_combined.bam > LT.vcf 
```

(note: I edited the freebayes-parallel script so that it would work...)

Freeybayes parallel takes about 12 hours

```{r, engine='bash', eval=FALSE}
scp whitney.plb.ucdavis.edu:2019/IreneSnps/LT.vcf.gz ../input/
```

## now analyze vcf in R

```{r}
library(tidyverse)
library(ggrepel)
```

get the vcf header
```{r}
vcf.header <- system("zgrep '#C' ../input/LT.vcf.gz",intern = TRUE) 
vcf.header
vcf.header <- vcf.header %>% 
  str_replace("#","") %>% #get rid of the pound sign
  str_split(pattern = "\t") %>% #split on the tabs
  magrittr::extract2(1)
vcf.header
```

get the data
```{r}
snps <- read_tsv("../input/LT.vcf.gz", na = c("","NA","."),comment="#",col_names = vcf.header) %>%
  select(-ID, -FILTER) # these are empty columns
snps
```

filter to keep snps where there is data from all samples
```{r}
snps <- snps %>% 
  filter({select(., matches("[0-9]")) %>% complete.cases() })
snps
```

```{r}
snps <- snps %>% 
  mutate(TOTAL_DEPTH= {str_extract(INFO, "DP=[0-9]*") %>% 
      str_remove("DP=") %>%
      as.numeric() }
  ) %>%
  filter(QUAL >=100,
         nchar(ALT)==1,
         TOTAL_DEPTH > quantile(TOTAL_DEPTH, 0.05),
         TOTAL_DEPTH < quantile(TOTAL_DEPTH, 0.95))
snps
```

unpack the information differnet samples:

```{r}
samples <- colnames(snps) %>% str_subset("^[0-9]")

for (s in samples) {
snps <- snps %>%
  separate(!!s, into=paste(s,c("gt","tot.depth","allele.depth","ref.depth","ref.qual","alt.depth","alt.qual","gt.lik"),sep="_"),
           sep=":", convert = TRUE)
}
snps
```

For the PCA we only need the genotype info
```{r}
gts <- snps %>%
  select(CHROM, POS, ends_with("_gt"))
gts
```

remove the 38LTR gample

```{r}
gts <- gts %>% select(-`38LTR_gt`)
```


convert this to numeric
```{r}
geno.numeric <- gts %>%
  select(-CHROM, -POS) %>%
  lapply(factor) %>% # convert charcters to "factors", where each category is internally represented as a number
  as.data.frame() %>% # reformat
  data.matrix() %>%#  convert to numeric
  t() 

colnames(geno.numeric) <- str_c(gts$CHROM, "_", gts$POS)

head(geno.numeric[,1:5],10)

dim(geno.numeric)
dim(gts)
```

get the principal components

```{r}
pca <- prcomp(geno.numeric)
summary(pca)
```

plot it
```{r}
plot.data <- pca$x %>%
  as.data.frame %>%
  rownames_to_column("sample") %>%
  mutate(response=str_extract(sample, "(LTR|LTWR)")) %>%
  gather(key="component", value="value",PC2:PC9)
  
plots <- map(sort(unique(plot.data$component)), function(x) {
  plot.data %>%
    filter(component==x) %>%
    ggplot(aes(x=PC1, y= value, label=sample, color=response)) +
      geom_point() + ylab(x)
  }
  )

plots  
```

## snps that contribute the most to PC1:

```{r}
loadings <- as.data.frame(pca$rotation) %>%
  rownames_to_column("snp") %>%
  select(PC1, snp) %>%
  arrange(desc(abs(PC1)))
loadings
```

```{r}
contributions <- loadings %>%
  separate(snp, into=c("contig", "pos"), sep="_") %>%
  group_by(contig) %>%
  summarize(abs.contribution = abs(sum(PC1)),
            contribution = sum(PC1),
            number_of_snps = n()) %>%
  arrange(desc(abs.contribution))
contributions
```

bring in seq lengths

```{r}
lengths <- read_csv("../input/Pinsaporeference1_lengths.csv", col_names = c("contig", "length"), skip=1) %>%
  mutate(contig = str_remove(contig, " .*"))
lengths
```

```{r}
contributions <- contributions %>% 
  left_join(lengths) %>%
  mutate(snps_per_100bp = round(number_of_snps / length * 100, 2)) %>%
  select(contig, contribution, length, number_of_snps, snps_per_100bp)
contributions
```

```{r}
write_csv(contributions, "../output/gene_contributions.csv.gz")
```

## make plot of PC1 vs PC2 for paper

```{r}
pc1pc2 <- pca$x %>%
  as.data.frame() %>%
  rownames_to_column("ID") %>%
  select(ID, PC1, PC2) %>%
  mutate(ID={str_replace(ID, "W", "N") %>%
       str_replace("RR", "R2") %>%
      str_remove_all("(X|_gt)") },
    response=ifelse(str_detect(ID,"N"), "no recovery", "recovery"))
pc1pc2
```

```{r}
pc1pc2 %>%
  ggplot(aes(x=PC1, y = PC2, label=ID, color=response)) +
  geom_point() +
  geom_text_repel(show.legend=FALSE, direction="y")
ggsave("../output/PCA.pdf")
```

Create a list of GCZN01054158.1 SNPs
```{r}
GCZN01054158.1.loadings <- loadings %>%
  filter(str_detect(snp, fixed("GCZN01054158.1"))) %>%
  separate(snp,into = c("contig", "position"), sep="_",convert = TRUE) %>%
  arrange(position)
GCZN01054158.1.loadings
```

```{r}
GCZN01054158.snpinfo <- left_join(GCZN01054158.1.loadings, snps, by=c("contig" = "CHROM", "position" = "POS"))
GCZN01054158.snpinfo
```

```{r}
write_csv(GCZN01054158.snpinfo, "../output/GCZN01054158.snpinfo.csv")
```

Create a list of SNPs in all genes with fixed differences

```{r}
get_snps <- function(contig, loadings=loadings) {
  loadings %>%
  filter(str_detect(snp, fixed("GCZN01054158.1"))) %>%
  separate(snp,into = c("contig", "position"), sep="_",convert = TRUE) %>%
  arrange(position)
GCZN01054158.1.loadings
GCZN01054158.snpinfo <- left_join(GCZN01054158.1.loadings, snps, by=c("contig" = "CHROM", "position" = "POS"))
GCZN01054158.snpinfo
}
```


```{r}
fixed_genes <- read_csv("../output/fixed_snps.csv")
fixed_genes
```

```{r}
fixed_genes_snps <- loadings %>% 
  separate(snp,into = c("contig", "position"), sep="_",convert = TRUE, remove = FALSE) %>%
  semi_join(fixed_genes, by="contig") %>%
  left_join(snps, by=c("contig" = "CHROM", "position" = "POS")) %>%
  arrange(contig, position)
fixed_genes_snps
```

```{r}
write_csv(fixed_genes_snps, "../output/fixed_genes_snps.csv")
```

