---
title: "MDS analysis"
output: html_notebook
---

Goal: count SNP difference between LTR and LTNR on a per gene basis


```{r}
library(tidyverse)
library(ggrepel)
```

get the vcf header
```{r}
vcf.header <- system("zgrep '#C' ../input/LT.vcf.gz",intern = TRUE) 
vcf.header
vcf.header <- vcf.header %>% 
  str_replace("#","") %>% #get rid of the pound sign
  str_split(pattern = "\t") %>% #split on the tabs
  magrittr::extract2(1)
vcf.header
```

get the data
```{r}
snps <- read_tsv("../input/LT.vcf.gz", na = c("","NA","."),comment="#",col_names = vcf.header) %>%
  select(-ID, -FILTER) # these are empty columns
snps
```

filter to keep snps where there is data from all samples
```{r}
snps <- snps %>% 
  filter({select(., matches("[0-9]")) %>% complete.cases() })
snps
```

```{r}
snps <- snps %>% 
  mutate(TOTAL_DEPTH= {str_extract(INFO, "DP=[0-9]*") %>% 
      str_remove("DP=") %>%
      as.numeric() }
  ) %>%
  filter(QUAL >=100,
         nchar(ALT)==1,
         TOTAL_DEPTH > quantile(TOTAL_DEPTH, 0.05),
         TOTAL_DEPTH < quantile(TOTAL_DEPTH, 0.95))
snps
```

unpack the information differnet samples:

```{r}
samples <- colnames(snps) %>% str_subset("^[0-9]")

for (s in samples) {
snps <- snps %>%
  separate(!!s, into=paste(s,c("gt","tot.depth","allele.depth","ref.depth","ref.qual","alt.depth","alt.qual","gt.lik"),sep="_"),
           sep=":", convert = TRUE)
}
snps
```


next steps

keep snps that are fixed differences between LTRR and LTNR

Count per gene

```{r}
snps.l <- snps %>% select(CHROM, POS, ends_with("_gt")) %>%
  gather(key="plant", value="gt", -CHROM, -POS) %>%
  mutate(respnse=str_extract(plant, "LT[NR]{1,2}"))
snps.l
```

